### PR TITLE
error handling during downgraded component bootstrap

### PR DESCRIPTION
**Pull request by FDIM**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

2 issues: 

- errors occurring in downgraded component constructor are unhandled and caught in zone.js 
- downgraded component is instantiated after scope is destroyed. In our project, since we are using lazy loading we are getting component factory not found errors.

Issue Number: #25613

## What is the new behavior?

- errors occurring in downgraded component constructor will be forwarded to angular.js $excetionHandler. We have missed some errors as they were not logged in production.
- downgraded component will not be created if scope was destroyed before angular was bootstrapped

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

### CODE REVIEW

**Comment by gkalpak** [comment]
Thx for the PR. I haven't looked into the problems this si solving, but from a quick look it seems reasonable :grin:
Can you add some tests demonstrating the problems being solved (and that they are indeed solved). I.e. we need some tests that would fail before and pass with the changes in the PR.

**Review by maclun — APPROVED** [review]
The code change looks good. It works well in our setup and resolved the issue.

**Comment by FDIM** [comment]
Hehe, I'll can try to reproduce them in some unit test, but it's not really easy as both of the fixes are for some edge cases. 

* $exceptionHandler is only about logging as some errors were not logged and swallowed by zone.js (#25613)
* checking for scope destroyed is an optimization in case some route resolver is taking a long time and user has already navigated away. For us this is an error because on route change we change 'current' lazy loaded module on ng-view (#17490) and thus component factory resolver fails to locate actual component

**Comment by FDIM** [comment]
Ready for review. @gkalpak I've managed to cover both cases in specs :) And I am glad that you didn't approve as I forgot to add exceptionHandler as a dependency in initial commit :D

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 1]
```diff
@@ -609,6 +609,41 @@ withEachNg1Version(() => {
            // Wait for the module to be bootstrapped.
            setTimeout(() => expect($injectorFromNg2).toBe($injectorFromNg1));
          }));
+
+      it('should forward errors to $exceptionHandler', (done) => {
+        @Component({selector: 'ng2', template: 'test'})
+        class TestComponent implements OnInit {
+          constructor(@Inject('Not Existing') _notExisting: string) {}
+          ngOnInit() {}
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: This is irrelevant. Please remove it.

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 2]
```diff
@@ -609,6 +609,41 @@ withEachNg1Version(() => {
...
+        }
+        @NgModule({
+          declarations: [TestComponent],
+          entryComponents: [TestComponent],
+          imports: [BrowserModule]
+        })
+        class Ng2Module {
+          constructor() {}
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: This is unnecessary. Please remove it.

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 3]
```diff
@@ -609,6 +609,41 @@ withEachNg1Version(() => {
            // Wait for the module to be bootstrapped.
            setTimeout(() => expect($injectorFromNg2).toBe($injectorFromNg1));
          }));
+
+      it('should forward errors to $exceptionHandler', (done) => {
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: Why not use `async()` as in the other tests?

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 4]
```diff
@@ -644,6 +644,45 @@ withEachNg1Version(() => {
           done();
         }, 0);
       });
+
+      it('should not create ng2 component instance if ng1 is destroyed ', (done) => {
+        let testComponentCreated = false;
+        @Component({selector: 'ng2', template: 'test'})
+        class TestComponent implements OnInit {
+          constructor() { testComponentCreated = true; }
+          ngOnInit() {}
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: This is irrelevant. Please remove it.

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 5]
```diff
@@ -609,6 +609,80 @@ withEachNg1Version(() => {
...
            setTimeout(() => expect($injectorFromNg2).toBe($injectorFromNg1));
          }));
+
+      it('should forward errors to $exceptionHandler', (done) => {
+        @Component({selector: 'ng2', template: 'test'})
+        class TestComponent implements OnInit {
+          constructor(@Inject('Not Existing') _notExisting: string) {}
+          ngOnInit() {}
+        }
+        @NgModule({
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: A few empty lines here and there never hurt anyone 😛
(Here and below.)

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 6]
```diff
@@ -609,6 +609,80 @@ withEachNg1Version(() => {
...
+          constructor() { testComponentCreated = true; }
+          ngOnInit() {}
+        }
+        @NgModule({
+          declarations: [TestComponent],
+          entryComponents: [TestComponent],
+          imports: [BrowserModule]
+        })
+        class Ng2Module {
+          constructor() {}
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: This is not needed. Please remove it.

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 7]
```diff
@@ -609,6 +609,80 @@ withEachNg1Version(() => {
...
+          ngDoBootstrap() {}
+        }
+        const element = html('<ng2 class="test"></ng2>');
+        const bootstrapFn = (extraProviders: StaticProvider[]) => {
+          return platformBrowserDynamic(extraProviders)
+              .bootstrapModule(Ng2Module)
+              .then((module) => {
+                // simulate scope destruction
+                angular.element(element).injector !().get($ROOT_SCOPE).$destroy();
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: Nit: You can use `$injectorFromNg1.get()` directly.

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts [hunk 8]
```diff
@@ -609,6 +609,80 @@ withEachNg1Version(() => {
...
+        @NgModule({
+          declarations: [TestComponent],
+          entryComponents: [TestComponent],
+          imports: [BrowserModule]
+        })
+        class Ng2Module {
+          constructor() {}
+          ngDoBootstrap() {}
+        }
+        const element = html('<ng2 class="test"></ng2>');
```
Comment (gkalpak) [inline packages/upgrade/test/static/integration/downgrade_module_spec.ts]: The `class` seems unnecessary.

File: packages/upgrade/src/common/downgrade_component.ts [hunk 9]
```diff
@@ -103,6 +104,12 @@ export function downgradeComponent(info: {
         }
 
         const doDowngrade = (injector: Injector) => {
+          // initial load can be async and the parent component may be destroyed before angular
+          // bootstrap e.g. slow resolve on some route and user navigation
```
Comment (gkalpak) [inline packages/upgrade/src/common/downgrade_component.ts]: I would rephrase this to something like:

> // Downgrading can happen asynchronously; e.g. an Angular module may need to be 
> // bootstrapped first. By the time we get here, the parent component may have been
> // destroyed, in which case there is no need to instantiate the downgraded component.

**Review by gkalpak — CHANGES_REQUESTED** [review]
I left some minor comments, but otherwise LGTM. Thx, @FDIM 💯
(If you make further changes, please add them as new [fixup commits][1] 🙏 This makes them much easier to review.)

[1]: https://git-scm.com/docs/git-commit#git-commit---fixupltcommitgt

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 4 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: done

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 3 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: done

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 1 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: done

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 5 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: done

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 6 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: done

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 7 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: Done. This is much nicer :)

File: packages/upgrade/test/static/integration/downgrade_module_spec.ts (hunk 8 above)
Comment (FDIM) [reply packages/upgrade/test/static/integration/downgrade_module_spec.ts]: removed

File: packages/upgrade/src/common/downgrade_component.ts (hunk 9 above)
Comment (FDIM) [reply packages/upgrade/src/common/downgrade_component.ts]: done

**Comment by FDIM** [comment]
@gkalpak PR should be good now

**Review by gkalpak — CHANGES_REQUESTED** [review]
LGTM 👍

I just think that [downgrade_component_spec.ts][1] might be a better place for these tests and we also need tests without `downgradeModule()`. Could you try moving the tests over to `downgrade_component_spec.ts` and add the same tests without `downgradeModule()`?

[1]: https://github.com/angular/angular/blob/23f2bf14aeed1705ab872f1e1a4e6d9e7581d881/packages/upgrade/test/static/integration/downgrade_component_spec.ts

**Comment by AndrewKushnir** [comment]
Closing this PR due to the lack of response. Please reopen if needed.
